src/dbn_struct.jl

```diff
@@ -0,0 +1,174 @@
+"""
+    DBNVar(x::Symbol)
+    DBNVar{x::Symbol}()
+
+Reference to a named node in the POMDP or MDP dynamic Bayesian network (DBN).
+
+Note that `gen(::DBNVar, m, depargs..., rng)` always takes an argument for each dependency, `gen(::DBNOut, m, s, a, rng)` only takes `s` and `a` arguments (the inputs to the entire DBN).
+
+`DBNVar` is a "value type". See the documentation of `Val` for more conceptual details about value types.
+"""
+struct DBNVar{name} end
+
+DBNVar(name::Symbol) = DBNVar{name}()
+
+"""
+    DBNOut(x::Symbol)
+    DBNOut{x::Symbol}()
+    DBNOut(::Symbol, ::Symbol,...)
+    DBNOut{x::NTuple{N, Symbol}}()
+
+Reference to one or more named nodes in the POMDP or MDP dynamic Bayesian network (DBN).
+
+Note that `gen(::DBNOut, m, s, a, rng)` always takes `s` and `a` arguments (the inputs to the entire DBN) while `gen(::DBNVar, m, depargs..., rng)` takes a variable number of arguments (one for each dependency).
+
+`DBNOut` is a "value type". See the documentation of `Val` for more conceptual details about value types.
+"""
+struct DBNOut{names} end
+
+DBNOut(name::Symbol) = DBNOut{name}()
+DBNOut(names...) = DBNOut{names}()
+
+struct DBNDef{N<:NamedTuple}
+    nodes::N
+    deps::NamedTuple # this could be a parameter in the future
+end
+
+# todo DBNDef structure
+
+node(d::DBNDef, name::Symbol) = d.nodes[name]
+deps(d::DBNDef, name::Symbol) = d.deps[name]
+nodenames(d::DBNDef) = keys(d.nodes)
+@generated function add_node(d::DBNDef, n::DBNVar{name}, node, deps) where name
```

Comment:
Just for my own understanding: why does this need to be a generated function?

Reply:
maybe there is a way to make it not @generated, but it seemed like the easiest way to create NamedTuples with `name` as the keys

Reply:
Can you document this function? It is not clear to me either why it needs to be generated

Reply:
You can create a named tuple with `name` as a key without `@generated` by writing `NamedTuple{tuple(name)}(tuple(node))`. I think this will cause less confusion